students/shawn_degraw/session09/mailroom_oo/donor_models.py

```diff
@@ -23,18 +22,131 @@ def name(self):
+    """
+
+    # Thank you letter templates
+    THANK_YOU_LETTER = "\n".join(("", "Dear {name},", "", "Thank you for your "
+                                  "generous donation of ${amount:.2f} to our "
+                                  "cause.", "Your donations help keep Python "
+                                  "great!", "", "Sincerely", "", "The Python "
+                                  "Project", ""))
+
+    GENERAL_DONATION_LETTER = "\n".join(("", "Dear {name},", "", "Thank you for "
+                                         "your generosity in supporting us with "
+                                         "${totaldonation:.2f} in donations.",
+                                         "We hope to have your continued support.",
+                                         "", "With great thanks", "", "The Python "
+                                         "Project"))
+
+    def __init__(self):
+        self.donor_collection = []
+
+    def donor_update(self, name, initial_donation=None):
+        """
+        Method to add donation if donor exists or
+        creates new Donor object if the donor does not exist.
+        Returns True if actions are successful.
+        Controls donation data and name data.
+
+        :param name: donor's name
+        :param initial_donation: initial donation value
+        """
+
+        # Validate the content of initial_donation and format it for storage
+        if not initial_donation:
+            return False
+        else:
+            try:
+                formatted_donation = int(float(initial_donation) * 100)
+            except ValueError:
+                return False
+
+        if formatted_donation < 1:
+            return False
+
+        # Storing name without extra spaces
+        formatted_name = " ".join(name.split())
+
+        # Create donor or update existing donor and print letter
+        donor = self.search_name(formatted_name)
+        if donor:
+            donor.add_donation(formatted_name, formatted_donation)
+            return True
+        else:
+            self.donor_collection.append(Donor(formatted_name, formatted_donation))
+            return True
+
+    def report_header(self):
+        """ Creates and returns report header as a string. """
+
+        return "\n{:<26}|{:^13}|{:^11}|{:^14}\n{}".format("Donor Name", "Total Given",
+                                                          "Num Gifts", "Average Gift", "-" * 66)
+
+    def create_report(self):
+        """ Generates report line by line and returns report as a string """
+
+        reportbody = self.report_header()
+        for donor in self.collect_data():
+            dobject = self.search_name(donor)
+            reportbody = "\n".join([reportbody, "{:<27}${:>11.2f} {:>11d}  ${:>12.2f}".format(dobject.name,
+                                    dobject.total_donations / 100, dobject.number_donations, dobject.average_donation / 100)])
+        return reportbody
```

Comment:
Minor: Better practice to construct a multi-line string output is to use a list:
```
lines = [self.report_header()]
for donr in self.collect_data():
    dobject = self.search_name(donor)
    lines.append("{:<27}${:>11.2f} {:>11d}  ${:>12.2f}".format(dobject.name., dobject.total_donations / 100, dobject.number_donations, dobject.average_donation / 100)
return '\n'.join(lines)
```

Reply:
Modified create_report to use a list instead of string. Easy fix and still returned a string so no change in my cli_main.py module.

Reply:
Saw it and looks good, though it could have been simpler with a list comprehension. No need to change further, though.